rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -344,6 +342,292 @@ impl From<RequestOrResponse> for CanisterInput {
+    /// responses.
+    fn is_empty(&self) -> bool {
+        self.pool.len() == 0 && self.expired_callbacks.is_empty() && self.shed_responses.is_empty()
+    }
+
+    /// Helper function for concisely validating the hard invariant that a canister
+    /// queue is either empty or has a non-stale reference at the front, by writing
+    /// `debug_assert_eq!(Ok(()), store.queue_front_not_stale(...)`.
+    ///
+    /// Time complexity: `O(log(n))`.
+    fn queue_front_not_stale<T>(
+        &self,
+        queue: &CanisterQueue<T>,
+        canister_id: &CanisterId,
+    ) -> Result<(), String>
+    where
+        MessageStoreImpl: MessageStore<T>,
+    {
+        if let Some(reference) = queue.peek() {
+            if self.is_stale(reference) {
+                return Err(format!(
+                    "Stale reference at the front of {:?} queue to/from {}",
+                    reference.context(),
+                    canister_id
+                ));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Defines context-specific (inbound / outbound) message store operations
+/// (lookup, removal, staleness check) for `MessageStoreImpl`.
+trait MessageStore<T> {
+    /// The type returned by `get()`: `&T` if the implementation actually holds
+    /// items of type `T`; or the type `T` if it has to be built on demand.
+    type TRef<'a>
+    where
+        Self: 'a;
+
+    /// Looks up the referenced item. Panics if the reference is stale.
+    fn get(&self, reference: message_pool::Reference<T>) -> Self::TRef<'_>;
+
+    /// Removes the referenced item. Panics if the reference is stale.
+    fn take(&mut self, reference: message_pool::Reference<T>) -> T;
+
+    /// Checks whether the given reference is stale (i.e. neither in the pool, nor
+    /// in one of the compact response maps iff inbound).
+    fn is_stale(&self, reference: message_pool::Reference<T>) -> bool;
+}
+
+impl MessageStore<CanisterInput> for MessageStoreImpl {
+    type TRef<'a> = CanisterInput;
+
+    fn get(&self, reference: InboundReference) -> CanisterInput {
+        assert_eq!(Context::Inbound, reference.context());
+
+        if let Some(msg) = self.pool.get(reference) {
+            debug_assert!(!self.expired_callbacks.contains_key(&reference));
+            debug_assert!(!self.shed_responses.contains_key(&reference));
+            return msg.clone().into();
+        } else if reference.class() == Class::BestEffort && reference.kind() == Kind::Response {
+            if let Some(callback_id) = self.expired_callbacks.get(&reference) {
+                debug_assert!(!self.shed_responses.contains_key(&reference));
+                return CanisterInput::DeadlineExpired(*callback_id);
+            } else if let Some(callback_id) = self.shed_responses.get(&reference) {
+                return CanisterInput::ResponseDropped(*callback_id);
+            }
+        }
+
+        panic!("stale reference at the front of input queue");
+    }
```

Comment:
Fair enough. How about a function
```Rust
fn is_best_effort_response(&self) -> bool {
   self.class() == Class::BestEffort && self.kind() == Kind::Response
}
```
or something like that? Maybe also something similar for `expired_callbacks` and `shed_responses` since both of these expressions also appear in the complicated logical expression @derlerd-dfinity mentioned below?